
#' @param intergenicRegions the intergenic regions generated by DERFINDER
#' @param coverageIntergenic
#' @param annotatedJunction a data.frame with the data generated

annotateERJunction <- function(intergenicRegions,coverageIntergenic,annotatedJunction)
{
    ##############################################################################
    ## this is the stage 1. annotation is done using the annotated split reads ###
    ##############################################################################

    ## first step is remove those junctions that are annotated from both ends. These are not adding any information and remove the unannotated junctions
    tmp.annotatedJunction <-  annotatedJunction[which((annotatedJunction$acceptor!="" | annotatedJunction$donor!="") & annotatedJunction$junction==""),]

    ## +- 1 in the coordinates to match the exon boundaries
    if(grepl("chr",as.character(annotatedJunction$chr[1])))
    {
        juncti.GR <- GRanges(c(as.character(tmp.annotatedJunction$chr),as.character(tmp.annotatedJunction$chr)),
                             IRanges(start=c(tmp.annotatedJunction$start-1,tmp.annotatedJunction$stop+1),
                                     end = c(tmp.annotatedJunction$start-1,tmp.annotatedJunction$stop+1)),
                             juncId=c(tmp.annotatedJunction$junId,tmp.annotatedJunction$junId))

    }else{
        juncti.GR <- GRanges(paste0("chr",c(as.character(tmp.annotatedJunction$chr),as.character(tmp.annotatedJunction$chr))),
                             IRanges(start=c(tmp.annotatedJunction$start-1,tmp.annotatedJunction$stop+1),
                                     end = c(tmp.annotatedJunction$start-1,tmp.annotatedJunction$stop+1)),
                             juncId=c(tmp.annotatedJunction$junId,tmp.annotatedJunction$junId))
    }

    regionJuncOve <- findOverlaps(intergenicRegions,juncti.GR,ignore.strand=FALSE)

    tmp.intergenicRegions <- as.data.frame(intergenicRegions)
    tmp.intergenicRegions[ ,c(colnames(tmp.annotatedJunction)[c(1,6:12)],"pred_strand","multi_junc","junER","distance","transcript","annotationType")] <- NA
    head(tmp.intergenicRegions)
    # seqnames   start     end width strand     value        area indexStart indexEnd cluster clusterL junId intronMotif inAnnotation countsSamples acceptor donor precBound pred_strand multi_junc junER distance transcript annotationType
    # 36    chr21 8228693 8228753    61      *   7.61245    464.3595       5337     5397      31       61    NA          NA           NA            NA       NA    NA        NA          NA         NA    NA       NA         NA             NA
    # 37    chr21 8234374 8234591   218      *  10.08276   2198.0420       5398     5615      32      218    NA          NA           NA            NA       NA    NA        NA          NA         NA    NA       NA         NA             NA
    # 38    chr21 8254013 8254736   724      * 433.28991 313701.8971       5616     6339      33      724    NA          NA           NA            NA       NA    NA        NA          NA         NA    NA       NA         NA             NA
    # 39    chr21 8255157 8255592   436      *  21.64786   9438.4674       6340     6775      34      436    NA          NA           NA            NA       NA    NA        NA          NA         NA    NA       NA         NA             NA
    # 40    chr21 8339142 8339156    15      * 240.90346   3613.5519       6776     6790      35       15    NA          NA           NA            NA       NA    NA        NA          NA         NA    NA       NA         NA             NA
    # 41    chr21 8354137 8354150    14      *  76.18537   1066.5951       6791     6804      36       55    NA          NA           NA            NA       NA    NA        NA          NA         NA    NA       NA         NA             NA

    head(regionJuncOve)
    # Hits object with 6 hits and 0 metadata columns:
    #   queryHits subjectHits
    # <integer>   <integer>
    #   [1]        17       55794
    # [2]        32      150504
    # [3]        52       55920
    # [4]        54       55921
    # [5]        67       55942
    # [6]        67       55943
    # -------
    #   queryLength: 332 / subjectLength: 189250

    ## adding the information on the junction id, selecting the junction id with maximum number of samples, if the region overlap with multiple junctions
    regionJuncOve.df <- as.data.frame(regionJuncOve)
    regionJuncOve.df <- cbind(regionJuncOve.df,countsSamples=tmp.annotatedJunction[match(mcols(juncti.GR[regionJuncOve.df$subjectHits])[,"juncId"],tmp.annotatedJunction$junId),"countsSamples"])
    setDT(regionJuncOve.df)
    regionJuncOve.df<- regionJuncOve.df [, .SD[which.max(countsSamples)], by=queryHits]
    setDF(regionJuncOve.df)

    tmp.intergenicRegions[regionJuncOve.df$queryHits,c(colnames( tmp.annotatedJunction)[c(1,6:12)],"pred_strand")] <- tmp.annotatedJunction[match(mcols(juncti.GR[regionJuncOve.df$subjectHits])[,"juncId"],tmp.annotatedJunction$junId),c(colnames( tmp.annotatedJunction)[c(1,6:12)],"strand")]

    tail(tmp.intergenicRegions)
    #      seqnames    start      end width strand      value         area indexStart indexEnd cluster clusterL  junId intronMotif inAnnotation countsSamples acceptor                           donor pred_strand multi_junc junER distance
    # 5257    chr21 46596867 46597232   366      *   8.620779 3.155205e+03    2266766  2267131    3576      568     NA          NA           NA            NA     <NA>                            <NA>        <NA>         NA    NA       NA
    # 5258    chr21 46598510 46602699  4190      * 334.707141 1.402423e+06    2267132  2271321    3577     4190 211908           2            1            98          ENST00000367071,ENST00000291700           -         NA    NA       NA
    # 5282    chr21 46665058 46665592   535      *   8.700233 4.654625e+03    2288259  2288793    3592     1298     NA          NA           NA            NA     <NA>                            <NA>        <NA>         NA    NA       NA
    # 5283    chr21 46668205 46668433   229      *   7.547700 1.728423e+03    2288794  2289022    3593      278     NA          NA           NA            NA     <NA>                            <NA>        <NA>         NA    NA       NA
    # 5284    chr21 46668474 46668482     9      *   6.456942 5.811248e+01    2289023  2289031    3593      278     NA          NA           NA            NA     <NA>                            <NA>        <NA>         NA    NA       NA
    # 5285    chr21 46692741 46692758    18      *   6.432018 1.157763e+02    2289032  2289049    3594       18     NA          NA           NA            NA     <NA>                            <NA>        <NA>         NA    NA

    ## adding the information in case regions have multiple id junctions overlappiong
    regionJuncOve.df <- as.data.frame(regionJuncOve)
    regionJuncOve.df <- cbind(regionJuncOve.df,junId=mcols(juncti.GR[regionJuncOve.df$subjectHits])[,"juncId"])
    setDT(regionJuncOve.df)
    regionJuncOve.df <- regionJuncOve.df[ , .(multiple_junc = paste(unique(junId), collapse=",")), by = queryHits]
    setDF(regionJuncOve.df)

    tmp.intergenicRegions[regionJuncOve.df$queryHits,"multi_junc"] <- regionJuncOve.df$multiple_junc
    tmp.intergenicRegions[regionJuncOve.df$queryHits,"annotationType"] <- 1

    tmp.intergenicRegions[!is.na(tmp.intergenicRegions$junId),"distance"] <- apply(abs(cbind(tmp.intergenicRegions[!is.na(tmp.intergenicRegions$junId),"start"] -
                                                                                                 annotatedJunction[match(tmp.intergenicRegions[!is.na(tmp.intergenicRegions$junId),"junId"],annotatedJunction$junId),"stop"],
                                                                                             tmp.intergenicRegions[!is.na(tmp.intergenicRegions$junId),"end"] -
                                                                                                 annotatedJunction[match(tmp.intergenicRegions[!is.na(tmp.intergenicRegions$junId),"junId"],annotatedJunction$junId),"stop"],
                                                                                             tmp.intergenicRegions[!is.na(tmp.intergenicRegions$junId),"start"] -
                                                                                                 annotatedJunction[match(tmp.intergenicRegions[!is.na(tmp.intergenicRegions$junId),"junId"],annotatedJunction$junId),"start"],
                                                                                             tmp.intergenicRegions[!is.na(tmp.intergenicRegions$junId),"end"] -
                                                                                                 annotatedJunction[match(tmp.intergenicRegions[!is.na(tmp.intergenicRegions$junId),"junId"],annotatedJunction$junId),"start"])),1,max)


    # add the transcript information
    regionJuncOve.df <- as.data.frame(regionJuncOve)
    regionJuncOve.df <- regionJuncOve.df <- cbind(regionJuncOve.df,tmp.annotatedJunction[match(mcols(juncti.GR[regionJuncOve.df$subjectHits])[,"juncId"],tmp.annotatedJunction$junId),c("acceptor","donor")])
    setDT(regionJuncOve.df)
    regionJuncOve.df <- regionJuncOve.df[ , .(transcipt = paste(unique(unlist(strsplit(as.character(c(acceptor,donor)),","))), collapse=",")), by = queryHits]
    setDF(regionJuncOve.df)
    tmp.intergenicRegions[regionJuncOve.df$queryHits,"transcript"] <- regionJuncOve.df$transcipt


    tail(tmp.intergenicRegions)
    # seqnames   start     end width strand     value        area indexStart indexEnd cluster clusterL  junId intronMotif inAnnotation countsSamples acceptor donor pred_strand           multi_junc junER distance
    # 36    chr21 8228693 8228753    61      *   7.61245    464.3595       5337     5397      31       61     NA          NA           NA            NA     <NA>  <NA>        <NA>                 <NA>    NA       NA
    # 37    chr21 8234374 8234591   218      *  10.08276   2198.0420       5398     5615      32      218     NA          NA           NA            NA     <NA>  <NA>        <NA>                 <NA>    NA       NA
    # 38    chr21 8254013 8254736   724      * 433.28991 313701.8971       5616     6339      33      724 208281           1            1           101                          + 208281,208283,208286    NA    44596
    # 39    chr21 8255157 8255592   436      *  21.64786   9438.4674       6340     6775      34      436     NA          NA           NA            NA     <NA>  <NA>        <NA>                 <NA>    NA       NA
    # 40    chr21 8339142 8339156    15      * 240.90346   3613.5519       6776     6790      35       15 208288           1            1           101                          +        208288,208291    NA   120833
    # 41    chr21 8354137 8354150    14      *  76.18537   1066.5951       6791     6804      36       55 208282           2            1           100                          -               208282    NA   143981

    ## make the data frame consistent
    tmp.intergenicRegions[tmp.intergenicRegions==""] <- NA

    #########################################################
    ## step 2. annotate using non-annotated split reads   ###
    #########################################################

    library(tidyverse)
    library(stringr)

    tmp.annotatedJunction <-  annotatedJunction[which(annotatedJunction$acceptor=="" | annotatedJunction$donor=="" | annotatedJunction$junction==""),]


    if(grepl("chr",as.character(annotatedJunction$chr[1])))
    {
        juncti.GR <- GRanges(c(as.character(tmp.annotatedJunction$chr),as.character(tmp.annotatedJunction$chr)),
                             IRanges(start=c(tmp.annotatedJunction$start-1,tmp.annotatedJunction$stop+1),
                                     end = c(tmp.annotatedJunction$start-1,tmp.annotatedJunction$stop+1)),
                             juncId=c(tmp.annotatedJunction$junId,tmp.annotatedJunction$junId))

    }else{
        juncti.GR <- GRanges(paste0("chr",c(as.character(tmp.annotatedJunction$chr),as.character(tmp.annotatedJunction$chr))),
                             IRanges(start=c(tmp.annotatedJunction$start-1,tmp.annotatedJunction$stop+1),
                                     end = c(tmp.annotatedJunction$start-1,tmp.annotatedJunction$stop+1)),
                             juncId=c(tmp.annotatedJunction$junId,tmp.annotatedJunction$junId))
    }

    regionJuncOve <- findOverlaps(intergenicRegions,juncti.GR,ignore.strand=FALSE)

    head(regionJuncOve)
    # Hits object with 6 hits and 0 metadata columns:
    #     queryHits subjectHits
    #     <integer>   <integer>
    # [1]         3      200184
    # [2]         3      200186
    # [3]         3      200189
    # [4]         5      200191
    # [5]         5      200194
    # [6]         6      200185
    # -------
    # queryLength: 332 / subjectLength: 258000

    ## adding the information on the junction id, selecting the junction id with maximum number of samples, if the region overlap with multiple junctions
    regionJuncOve.df <- as.data.frame(regionJuncOve)
    regionJuncOve.df <- cbind(regionJuncOve.df,junId=mcols(juncti.GR[regionJuncOve.df$subjectHits])[,"juncId"])
    ## I remove those indexes that are already annotated
    regionJuncOve.df <- regionJuncOve.df[!regionJuncOve.df$queryHits%in%which(!is.na(tmp.intergenicRegions$junId)),]

    regionJuncOve.df <- cbind(regionJuncOve.df,countsSamples=tmp.annotatedJunction[match(mcols(juncti.GR[regionJuncOve.df$subjectHits])[,"juncId"],tmp.annotatedJunction$junId),"countsSamples"])
    setDT(regionJuncOve.df)
    regionJuncOve.df<- regionJuncOve.df [, .SD[which.max(countsSamples)], by=queryHits]
    setDF(regionJuncOve.df)

    ## link regions using the split reads
    # regionJuncOve.df <- as.data.frame(regionJuncOve)
    # regionJuncOve.df <- cbind(regionJuncOve.df,junId=mcols(juncti.GR[regionJuncOve.df$subjectHits])[,"juncId"])
    # regionJuncOve.df <- regionJuncOve.df[!regionJuncOve.df$queryHits%in%which(!is.na(tmp.intergenicRegions$junId)),]

    regionJuncOve.df.tmp <-
        regionJuncOve.df %>%
        mutate(queryhits_junId = str_c(junId, "_", queryHits))

    regionJuncOve.df <- regionJuncOve.df %>%
        inner_join(regionJuncOve.df.tmp, by = "junId") %>%
        arrange(junId) %>%
        mutate(to_keep = ifelse(duplicated(queryhits_junId, fromLast = T) | duplicated(queryhits_junId, fromLast = F),
                                ifelse(queryHits.x == queryHits.y, F, T), T)) %>%
        filter(to_keep == T) %>%
        mutate(queryHits_dup = ifelse(queryHits.x == queryHits.y, NA, queryHits.y)) %>%
        dplyr::select(queryHits = queryHits.x, subjectHits.x, junId, queryHits_dup)

    regionJuncOve.df <- regionJuncOve.df[!is.na(regionJuncOve.df$queryHits_dup),]


    ## add inforamtion about the junction

    ## add information about the relationship between two Expressed regions
    tmp.intergenicRegions[regionJuncOve.df$queryHits,"junER"] <- regionJuncOve.df$queryHits_dup

    tmp.intergenicRegions[regionJuncOve.df$queryHits,"annotationType"] <- 2

    # add the transcript information and distance this is remove for the moment, I want to speak with Mina first
    # tmp.intergenicRegions[regionJuncOve.df$queryHits,c(colnames( tmp.annotatedJunction)[c(1,6:10,12)],"pred_strand")] <- tmp.annotatedJunction[match(mcols(juncti.GR[regionJuncOve.df$subjectHits])[,"juncId"],tmp.annotatedJunction$junId),c(colnames( tmp.annotatedJunction)[c(1,6:10,12)],"strand")]
    # regionJuncOve.df <- as.data.frame(regionJuncOve)
    # regionJuncOve.df <- cbind(regionJuncOve.df,tmp.annotatedJunction[match(mcols(juncti.GR[regionJuncOve.df$subjectHits])[,"juncId"],tmp.annotatedJunction$junId),c("acceptor","donor")])
    # regionJuncOve.df <- regionJuncOve.df[!regionJuncOve.df$queryHits%in%which(!is.na(tmp.intergenicRegions$junId)),]
    # setDT(regionJuncOve.df)
    # regionJuncOve.df <- regionJuncOve.df[ , .(transcipt = paste(unique(unlist(strsplit(as.character(c(acceptor,donor)),","))), collapse=",")), by = queryHits]
    # setDF(regionJuncOve.df)
    # tmp.intergenicRegions[regionJuncOve.df$queryHits,"transcript"] <- regionJuncOve.df$transcipt
    #
    # ## calculate the distance using the minimum of all the distances between start and stop of the regions
    # tmp.intergenicRegions[!is.na(tmp.intergenicRegions$junER),"distance"] <- apply(abs(cbind(tmp.intergenicRegions[!is.na(tmp.intergenicRegions$junER),"start"] -
    #                                                                                            tmp.intergenicRegions[tmp.intergenicRegions[!is.na(tmp.intergenicRegions$junER),"junER"],"end"],
    #                                                                                          tmp.intergenicRegions[!is.na(tmp.intergenicRegions$junER),"end"] -
    #                                                                                            tmp.intergenicRegions[tmp.intergenicRegions[!is.na(tmp.intergenicRegions$junER),"junER"],"start"],
    #                                                                                          tmp.intergenicRegions[!is.na(tmp.intergenicRegions$junER),"start"] -
    #                                                                                            tmp.intergenicRegions[tmp.intergenicRegions[!is.na(tmp.intergenicRegions$junER),"junER"],"end"],
    #                                                                                          tmp.intergenicRegions[!is.na(tmp.intergenicRegions$junER),"end"] -
    #                                                                                            tmp.intergenicRegions[tmp.intergenicRegions[!is.na(tmp.intergenicRegions$junER),"junER"],"start"])),1,min)

    rm(regionJuncOve)

    ## make the data frame consistent
    tmp.intergenicRegions[tmp.intergenicRegions==""] <- NA

    ########################################################################################
    ## step 3. annotate regions based on the nearest gene that have junction annotated  ####
    ########################################################################################

    tmp.annotatedJunction <-  annotatedJunction[which(annotatedJunction$acceptor!="" | annotatedJunction$donor!="" |annotatedJunction$junction!=""),]

    ## +- 1 in the coordinates to match the exon boundaries
    if(grepl("chr",as.character(annotatedJunction$chr[1])))
    {
        juncti.GR <- GRanges(c(as.character(tmp.annotatedJunction$chr),as.character(tmp.annotatedJunction$chr)),
                             IRanges(start=c(tmp.annotatedJunction$start-1,tmp.annotatedJunction$stop+1),
                                     end = c(tmp.annotatedJunction$start-1,tmp.annotatedJunction$stop+1)),
                             juncId=c(tmp.annotatedJunction$junId,tmp.annotatedJunction$junId))

    }else{
        juncti.GR <- GRanges(paste0("chr",c(as.character(tmp.annotatedJunction$chr),as.character(tmp.annotatedJunction$chr))),
                             IRanges(start=c(tmp.annotatedJunction$start-1,tmp.annotatedJunction$stop+1),
                                     end = c(tmp.annotatedJunction$start-1,tmp.annotatedJunction$stop+1)),
                             juncId=c(tmp.annotatedJunction$junId,tmp.annotatedJunction$junId))
    }

    ## calculate the distance to nearest junction, junction which is annotated with a gene
    distNearest <- as.data.frame(distanceToNearest(intergenicRegions,juncti.GR,ignore.strand=FALSE))
    ## add the juncId
    distNearest <- cbind(distNearest,junId=mcols(juncti.GR[distNearest$subjectHits])[,"juncId"])
    ## add the distance for those records were we already have added the distance based on the overlapping junction

    distNearest <- distNearest[!distNearest$queryHits%in%which(!is.na(tmp.intergenicRegions$junId)),]
    head(distNearest)
    # queryHits subjectHits distance  junId
    # 1         1      189013   177302 208302
    # 2         2      189013   171464 208302
    # 3         3      189013   151319 208302
    # 4         4      189013   150463 208302
    # 5         5      189013    66899 208302
    # 6         6      189013    51905 208302

    distNearest <- cbind(distNearest,tmp.annotatedJunction[match(mcols(juncti.GR[distNearest$subjectHits])[,"juncId"],tmp.annotatedJunction$junId),c("strand","intronMotif","inAnnotation","countsSamples","acceptor","donor","precBound","junction")])

    tmp.intergenicRegions[distNearest$queryHits,c(colnames( tmp.annotatedJunction)[c(1,6:10,12)],"pred_strand","distance")] <-
        distNearest[,c("junId","intronMotif","inAnnotation","countsSamples","acceptor","donor","precBound","strand","distance")]

    ## first speak to Mina
    ##tmp.intergenicRegions[distNearest$queryHits,"annotationType"] <- 3
    ## otherwise leave this
    tmp.intergenicRegions[distNearest[distNearest$queryHits%in%which(is.na(tmp.intergenicRegions$annotationType)) ,"queryHits"],"annotationType"] <- 3


    table(tmp.intergenicRegions$annotationType)

    setDT(distNearest)
    distNearest <- distNearest[ , .(transcipt = paste(unique(unlist(strsplit(as.character(c(acceptor,donor,junction)),","))), collapse=",")), by = queryHits]
    setDF(distNearest)

    tmp.intergenicRegions[distNearest$queryHits,"transcript"] <- distNearest$transcipt
}
